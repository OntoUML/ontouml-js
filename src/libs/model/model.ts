import {
  IElement,
  IContainer,
  IContainedElement,
} from '@libs/model/interfaces';
import memoizee from 'memoizee';



/**
 * Class that represents a Model according to `ontouml-schema`.
 *
 * @author Claudenir Fonseca
 * @author Lucas Bassetti
 */
export class Model implements IElement, IContainer {
  type: 'Model';
  id: string;
  name: string | null;
  authors: string[] | null;
  elements: IContainedElement[] | null;

  constructor(
    id: string,
    enableHash = false,
    name?: string,
    authors?: string[],
    elements?: IContainedElement[],
  ) {
    this.id = id;
    this.name = name;
    this.authors = authors;
    this.elements = elements;

    if (enableHash) {
      this.getAllElements = memoizee(this.getAllElements);
    }
  }

  getAllElements(): IContainedElement {
    var allElements = [...this.elements];

    this.elements.forEach(element => {
        allElements.push(element);
        if((element as unknown as IContainer).getAllElements) {
            const innerElements = element.getAllElements();
            if()
        }
    });

    return allElements;
  }
}
