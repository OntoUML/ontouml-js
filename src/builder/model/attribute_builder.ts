import {
  Cardinality,
  CardinalityValues,
  Class,
  Classifier,
  ModelElement,
  Property,
  BEGIN,
  END,
  DecoratableBuilder,
  PropertyStereotype
} from '../..';

export class AttributeBuilder extends DecoratableBuilder<
  AttributeBuilder,
  PropertyStereotype
> {
  protected override element?: Property;
  protected override readonly _container: Class;
  private _propertyType?: Classifier<any, any>;
  private _subsettedProperties: Set<Property> = new Set();
  private _redefinedProperties: Set<Property> = new Set();
  private _cardinality: Cardinality = new Cardinality(
    CardinalityValues.ZERO_TO_MANY
  );
  private _isOrdered: boolean = false;
  private _isReadOnly: boolean = false;

  constructor(c: Class) {
    super(c.project!);
    this._container = c;
  }

  /**
   * Builds an instance of {@link Property} with the parameters passed to the builder. **WARNING:** the ordering in which methods are evoked may affect the resulting object. When no methods are evoked, the created class has the following defaults:
   * - `id: "randomly-generated-id",`
   * - `created: new Date(),`
   * - `isDerived: false,`
   * - `isReadOnly: false,`
   * - `isOrdered: false,`
   * - `cardinality: "0..*"`
   */
  override build(): Property {
    this.element = new Property(this.project);

    super.build();

    this.element.propertyType = this._propertyType;
    this._subsettedProperties.forEach(
      p => this.element?.addSubsettedProperty(p)
    );
    this._redefinedProperties.forEach(
      p => this.element?.addRedefinedProperty(p)
    );
    this.element.cardinality = this._cardinality;
    this.element.isOrdered = this._isOrdered;
    this.element.isReadOnly = this._isReadOnly;

    this._container.addAttribute(this.element);
    return this.element;
  }

  override container(_: ModelElement): AttributeBuilder {
    throw new Error('Container already set on constructor.');
  }

  type(c: Class): AttributeBuilder {
    this._propertyType = c;
    return this;
  }

  ordered(): AttributeBuilder {
    this._isOrdered = true;
    return this;
  }

  unordered(): AttributeBuilder {
    this._isOrdered = false;
    return this;
  }

  readOnly(): AttributeBuilder {
    this._isReadOnly = true;
    return this;
  }

  writable(): AttributeBuilder {
    this._isReadOnly = false;
    return this;
  }

  subsets(p: Property): AttributeBuilder {
    this._subsettedProperties.add(p);
    return this;
  }

  redefines(p: Property): AttributeBuilder {
    this._redefinedProperties.add(p);
    return this;
  }

  cardinality(value: string): AttributeBuilder {
    this._cardinality = new Cardinality(value);
    return this;
  }

  optional(): AttributeBuilder {
    return this.cardinality(CardinalityValues.ZERO_TO_ONE);
  }

  many(): AttributeBuilder {
    return this.cardinality(CardinalityValues.MANY);
  }

  one(): AttributeBuilder {
    return this.cardinality(CardinalityValues.ONE);
  }

  some(): AttributeBuilder {
    return this.cardinality(CardinalityValues.ONE_TO_MANY);
  }

  /**
   * Sets the stereotype field and set default values in case of a known PropertyStereotype.
   */
  override stereotype(stereotype: string): AttributeBuilder {
    switch (stereotype) {
      case BEGIN:
        return this.begin();
      case END:
        return this.end();
    }

    return super.stereotype(stereotype);
  }

  /**
   * Sets the stereotype field to `«begin»` as well as the following default values:
   * - `isReadOnly = true`
   * - `cardinality = "1"`
   */
  begin(): AttributeBuilder {
    this._stereotype = BEGIN;
    this.readOnly();
    this.one();
    return this;
  }

  /**
   * Sets the stereotype field to `«end»` as well as the following default values:
   * - `isReadOnly = true`
   * - `cardinality = "1"`
   */
  end(): AttributeBuilder {
    this._stereotype = END;
    this.readOnly();
    this.one();
    return this;
  }
}
