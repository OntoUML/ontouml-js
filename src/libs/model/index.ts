import Ajv from 'ajv';
import memoizee from 'memoizee';
import { OntoUMLType } from '@constants/.';
import { IPackage, IElement, IRelation, IClass, IReference } from '@types';
import functions from '@libs/model/functions';

/**
 * Utility class for the manipulation of models in OntoUML2 in conformance to the `ontouml-schema` definition.
 *
 * @author Claudenir Fonseca
 * @author Lucas Bassetti
 *
 * @todo Create function to access this.sourceElements that throws an exception when the element is not found.
 */
export class ModelManager {
  model: IPackage;
  allElements: { [key: string]: IElement };

  constructor(model: IPackage) {
    // let validator = new Ajv().compile(schemas.getSchema(schemas.ONTOUML_2));
    const schema = require('@schemas/versions/ontouml-v1.0.schema');
    let validator = new Ajv().compile(schema);
    let isValid = validator(model);

    console.log('Checking validity');

    if (!isValid) {
      throw {
        message: 'Invalid model input.',
        errors: validator.errors,
      };
    }

    // Enabling memoization
    console.log('Enabling general memoization.');
    this.model = model;
    this.allElements = {};
    this.getElements = memoizee(this.getElements);
    this.getElementById = memoizee(this.getElementById);

    // Creating elements
    console.log('Creating elements.');
    this.getElements().forEach((element: IElement) => {
      this.allElements[element.id] = element;
      this.injectFunctions(element);
    });

    // Resolving references
    console.log('Resolving references.');
    this.getElements().forEach((element: IElement) => {
      this.replaceReferences(element);
    });
  }

  /**
   * Function that returns an array of contained IElement objects and updates the container IReference of all elements along the way.
   *
   * @param packageElement Optional package parameter that, when passed, makes the function return the elements contained within the package rather than the model itself.
   *
   * @returns Array of IElement objects contained within the model or the package parameter.
   */
  getElements(current?: IPackage): IElement[] {
    current = current ? current : this.model;
    current.container = null;

    if (!current.contents) {
      return [];
    }

    let contents = current.contents || [];

    current.contents.forEach((element: IElement) => {
      if (element.type === OntoUMLType.PACKAGE_TYPE) {
        contents = [...contents, ...this.getElements(element as IPackage)];
      } else if (element.type === OntoUMLType.RELATION_TYPE) {
        const relation = element as IRelation;

        contents = [
          ...contents,
          ...(relation.properties ? relation.properties : []),
        ];
      } else if (element.type === OntoUMLType.CLASS_TYPE) {
        const _class = element as IClass;

        contents = [
          ...contents,
          ...(_class.properties ? _class.properties : []),
          ...(_class.literals ? _class.literals : []),
        ];
      }

      element.container = { type: current.type, id: current.id };
    });

    return contents;
  }

  /**
   * Returns an IElement according to its id.
   *
   * @param elementId - Desired element's id.
   */
  getElementById(elementId: string): any {
    return this.getElements().find(element => element.id === elementId);
  }

  checkInstanceOfIContainer(element: IElement): boolean {
    return [
      OntoUMLType.PACKAGE_TYPE,
      OntoUMLType.CLASS_TYPE,
      OntoUMLType.RELATION_TYPE,
    ].includes(element.type);
  }

  checkInstanceOfIDecoratable(element: IElement): boolean {
    return [
      OntoUMLType.PROPERTY_TYPE,
      OntoUMLType.CLASS_TYPE,
      OntoUMLType.RELATION_TYPE,
    ].includes(element.type);
  }

  checkInstanceOfIClassifier(element: IElement): boolean {
    return [OntoUMLType.CLASS_TYPE, OntoUMLType.RELATION_TYPE].includes(
      element.type,
    );
  }

  injectFunctions(element: IElement): void {
    Object.keys(functions.IElement_functions).forEach(
      (functionName: string) => {
        element[functionName] = functions.IElement_functions[functionName];
      },
    );

    if (this.checkInstanceOfIContainer(element)) {
      Object.keys(functions.IContainer_functions).forEach(
        (functionName: string) => {
          element[functionName] = functions.IContainer_functions[functionName];
        },
      );
    }

    if (this.checkInstanceOfIClassifier(element)) {
      Object.keys(functions.IClassifier_functions).forEach(
        (functionName: string) => {
          element[functionName] = functions.IClassifier_functions[functionName];
        },
      );
    }
  }

  ejectFunctions(element: IElement): void {
    Object.keys(functions.IElement_functions).forEach(
      (functionName: string) => {
        element[functionName] = undefined;
      },
    );

    if (this.checkInstanceOfIContainer(element)) {
      Object.keys(functions.IContainer_functions).forEach(
        (functionName: string) => {
          element[functionName] = undefined;
        },
      );
    }

    if (this.checkInstanceOfIClassifier(element)) {
      Object.keys(functions.IClassifier_functions).forEach(
        (functionName: string) => {
          element[functionName] = undefined;
        },
      );
    }
  }

  replaceReferences(element: IElement): void {
    Object.keys(element).forEach((elementKey: string) => {
      if (JSON.stringify(Object.keys(element).sort()) === '["id","type"]') {
        element[elementKey] = this.getElementByReference(element[elementKey]);
      }
    });
  }

  getElementByReference(reference: IReference): IElement {
    const element: IElement = this.allElements[reference.id];

    if (!element || element.type !== reference.type) {
      throw {
        message: 'Bad reference.',
        reference: reference,
      };
    }

    return element;
  }

  // /**
  //  * Returns an IElement with PENDING references
  //  *
  //  * @param source - element serialized according to `ontouml-schema`.
  //  */
  // createIElement(source: any, enableMemoization = false): IElement {
  //   let element: any = {};

  //   // Checks for IElement
  //   if (
  //     [
  //       OntoUMLType.PACKAGE_TYPE,
  //       OntoUMLType.CLASS_TYPE,
  //       OntoUMLType.RELATION_TYPE,
  //       OntoUMLType.PROPERTY_TYPE,
  //       OntoUMLType.GENERALIZATION_TYPE,
  //       OntoUMLType.GENERALIZATION_SET_TYPE,
  //     ].includes(source.type)
  //   ) {
  //     element.type = source.type;
  //     element.id = source.id;
  //     element.name = source.name;
  //     element.description = source.description ? source.description : null;
  //     element.propertyAssignments = null;
  //     element.container = null;
  //     element = { ...element, ...functions.IElement_functions };
  //   } else {
  //     throw {
  //       name: 'Trying to create an IElement out of a non-recongnized source.',
  //       error: source,
  //     };
  //   }

  //   // Checks for IContainer
  //   if (
  //     [
  //       OntoUMLType.PACKAGE_TYPE,
  //       OntoUMLType.CLASS_TYPE,
  //       OntoUMLType.RELATION_TYPE,
  //     ].includes(source.type)
  //   ) {
  //     element = { ...element, ...functions.IContainer_functions };
  //   }

  //   // Checks for IDecoratable
  //   if (
  //     [
  //       OntoUMLType.CLASS_TYPE,
  //       OntoUMLType.RELATION_TYPE,
  //       OntoUMLType.PROPERTY_TYPE,
  //     ].includes(source.type)
  //   ) {
  //     element.stereotypes = source.stereotypes;
  //   }

  //   // Checks for IClassifier
  //   if (
  //     [OntoUMLType.CLASS_TYPE, OntoUMLType.RELATION_TYPE].includes(source.type)
  //   ) {
  //     element.properties = source.properties;
  //     element.isAbstract = source.isAbstract;
  //     element.isDerived = source.isDerived;
  //     element = { ...element, ...functions.IClassifier_functions };
  //   }

  //   // Checks for remaining interfaces
  //   switch (source.type) {
  //     case OntoUMLType.PACKAGE_TYPE:
  //       element.elements = null;
  //       return element as IPackage;

  //     case OntoUMLType.CLASS_TYPE:
  //       return element as IClass;

  //     case OntoUMLType.RELATION_TYPE:
  //       return element as IRelation;

  //     case OntoUMLType.GENERALIZATION_TYPE:
  //       element.general = null;
  //       element.specific = null;
  //       return element as IGeneralization;

  //     case OntoUMLType.GENERALIZATION_SET_TYPE:
  //       element.isDisjoint = source.isDisjoint;
  //       element.isComplete = source.isComplete;
  //       element.categorizer = null;
  //       element.generalizations = null;
  //       return element as IGeneralizationSet;

  //     case OntoUMLType.PROPERTY_TYPE:
  //       element.cardinality = source.cardinality;
  //       element.aggregationKind = source.aggregationKind;
  //       element.isDerived = source.isDerived;
  //       element.isOrdered = source.isOrdered;
  //       element.isReadOnly = source.isReadOnly;
  //       element.propertyType = null;
  //       element.subsettedProperties = null;
  //       element.redefinedProperties = null;
  //       return element as IProperty;
  //   }

  //   return null;
  // }

  // /**
  //  * Resolves pending references of created IElement.
  //  *
  //  * @param generatedElement - IElement with PENDING references
  //  */
  // resolveIElement(generatedElement: IElement): void {
  //   const sourceElement: any = this.sourceElements[generatedElement.id];

  //   // Checks for IElement
  //   if (
  //     [
  //       OntoUMLType.PACKAGE_TYPE,
  //       OntoUMLType.CLASS_TYPE,
  //       OntoUMLType.RELATION_TYPE,
  //       OntoUMLType.PROPERTY_TYPE,
  //       OntoUMLType.GENERALIZATION_TYPE,
  //       OntoUMLType.GENERALIZATION_SET_TYPE,
  //     ].includes(generatedElement.type)
  //   ) {
  //     (generatedElement as IElement).container = this.createdElements[
  //       sourceElement.container
  //     ] as IContainer;
  //   }

  //   // Checks for IClassifier
  //   if (
  //     [OntoUMLType.CLASS_TYPE, OntoUMLType.RELATION_TYPE].includes(
  //       generatedElement.type,
  //     ) &&
  //     sourceElement.elements
  //   ) {
  //     sourceElement.elements.forEach(
  //       item =>
  //         ((generatedElement as IClassifier).properties = [
  //           ...(generatedElement as IClassifier).properties,
  //           this.createdElements[item.id],
  //         ]),
  //     );
  //   }

  //   // Checks for remaining interfaces
  //   switch (generatedElement.type) {
  //     case OntoUMLType.PACKAGE_TYPE:
  //       if (sourceElement.elements && (generatedElement as IPackage).contents) {
  //         sourceElement.elements.forEach(
  //           item =>
  //             ((generatedElement as IPackage).contents = [
  //               ...(generatedElement as IPackage).contents,
  //               this.createdElements[item.id],
  //             ]),
  //         );
  //       }
  //       break;

  //     case OntoUMLType.CLASS_TYPE:
  //       break;

  //     case OntoUMLType.RELATION_TYPE:
  //       break;

  //     case OntoUMLType.GENERALIZATION_TYPE:
  //       (generatedElement as IGeneralization).general = this.createdElements[
  //         sourceElement.general.id
  //       ] as IClassifier;
  //       (generatedElement as IGeneralization).specific = this.createdElements[
  //         sourceElement.specific.id
  //       ] as IClassifier;
  //       break;

  //     case OntoUMLType.GENERALIZATION_SET_TYPE:
  //       (generatedElement as IGeneralizationSet).categorizer = this
  //         .createdElements[sourceElement.categorizer.id] as IClass;
  //       sourceElement.generalizations.forEach(
  //         item =>
  //           ((generatedElement as IGeneralizationSet).generalizations = [
  //             ...(generatedElement as IGeneralizationSet).generalizations,
  //             this.createdElements[item.id],
  //           ]),
  //       );
  //       break;

  //     case OntoUMLType.PROPERTY_TYPE:
  //       (generatedElement as IProperty).propertyType = this.createdElements[
  //         sourceElement.propertyType.id
  //       ] as IClassifier;
  //       if (sourceElement.subsettedProperties) {
  //         sourceElement.subsettedProperties.forEach(
  //           item =>
  //             ((generatedElement as IProperty).subsettedProperties = [
  //               ...(generatedElement as IProperty).subsettedProperties,
  //               this.createdElements[item.id],
  //             ]),
  //         );
  //       }
  //       if (sourceElement.redefinedProperties) {
  //         sourceElement.redefinedProperties.forEach(
  //           item =>
  //             ((generatedElement as IProperty).redefinedProperties = [
  //               ...(generatedElement as IProperty).redefinedProperties,
  //               this.createdElements[item.id],
  //             ]),
  //         );
  //       }
  //       break;
  //   }

  //   return null;
  // }
}
