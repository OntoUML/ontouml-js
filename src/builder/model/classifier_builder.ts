import { Classifier } from '../../model/classifier';
import { Stereotype } from '../../model/stereotypes';
import { DecoratableBuilder } from './decoratable_builder';

export abstract class ClassifierBuilder<
  B extends ClassifierBuilder<B, S>,
  S extends Stereotype
> extends DecoratableBuilder<B, S> {
  protected override element?: Classifier<any, any>;
  protected _isAbstract: boolean = false;

  override build(): Classifier<any, any> {
    super.build();

    this.assertElement();
    this.element!.isAbstract = this._isAbstract;

    return this.element!;
  }

  abstract(): B {
    this._isAbstract = true;
    return this as unknown as B;
  }

  concrete(): B {
    this._isAbstract = false;
    return this as unknown as B;
  }
}
