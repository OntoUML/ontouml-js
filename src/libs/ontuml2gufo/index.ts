import { N3Writer } from 'n3';
import { ModelManager } from '@libs/model';
import {
  IClass,
  IPackage,
  IGeneralizationSet,
  IGeneralization,
  IRelation,
  IOntoUML2GUFOOptions,
} from '@types';
import { OntoUMLType } from '@constants/.';
import { transformDisjointClasses, transformClassesByStereotype } from './class_functions';
import { transformRelations } from './relation_functions';
import { getURI, getPrefixes } from './helper_functions';
import URIManager from './uri_manager';

const N3 = require('n3');
const { DataFactory } = N3;
const { namedNode, quad } = DataFactory;

/**
 * Utility class for transform OntoUML models in OWL using the gUFO ontology
 *
 * @author Claudenir Fonseca
 * @author Lucas Bassetti
 */
export class OntoUML2GUFO {
  model: IPackage;

  constructor(model: ModelManager) {
    this.model = model.rootPackage;
  }

  async transformOntoUML2GUFO(options: IOntoUML2GUFOOptions): Promise<string> {
    const { baseIRI, format } = options;

    options.uriManager = new URIManager();

    const packages = this.model.getAllContentsByType([OntoUMLType.PACKAGE_TYPE]) as IPackage[];
    const prefixes = await getPrefixes(packages, options);

    const writer = new N3.Writer({
      format: format || 'Turtle',
      prefixes: {
        ...prefixes,
        gufo: 'http://purl.org/nemo/gufo#',
        rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
        rdfs: 'http://www.w3.org/2000/01/rdf-schema#',
        owl: 'http://www.w3.org/2002/07/owl#',
        xsd: 'http://www.w3.org/2001/XMLSchema#',
      },
    });

    writer.addQuads([
      quad(namedNode(baseIRI), namedNode('rdf:type'), namedNode('owl:Ontology')),
      quad(namedNode(baseIRI), namedNode('owl:imports'), namedNode('gufo:')),
    ]);

    await Promise.all([
      this.transformOntoUMLClasses2GUFO(writer, options),
      this.transformGeneralizationSets(writer, options),
      this.transformOntoUMLRelations2GUFO(writer, options),
    ]);

    return await new Promise<string>((resolve: (result: string) => null) => {
      writer.end((error: any, result: string) => {
        if (error) {
          console.log(error);
        }

        resolve(result);
      });
    });
  }

  /**
   * Main method to transform OntoUML classes in gUFO. The method will be responsable to run different class transformations.
   */
  async transformOntoUMLClasses2GUFO(writer: N3Writer, options: IOntoUML2GUFOOptions) {
    const classes = this.model.getAllContentsByType([OntoUMLType.CLASS_TYPE]) as IClass[];

    await Promise.all([
      transformDisjointClasses(writer, classes, options),
      transformClassesByStereotype(writer, classes, options),
    ]);

    return true;
  }

  /**
   * Main method to transform OntoUML relations in gUFO. The method will be responsable to run different relations transformations.
   */
  async transformOntoUMLRelations2GUFO(writer: N3Writer, options: IOntoUML2GUFOOptions) {
    const relations = this.model.getAllContentsByType([OntoUMLType.RELATION_TYPE]) as IRelation[];

    await transformRelations(writer, relations, options);

    return true;
  }

  /**
   * The method will be responsable to parse all properties (disjoint, complete) of generalization sets.
   */
  async transformGeneralizationSets(writer: N3Writer, options: IOntoUML2GUFOOptions) {
    const generalizationSets = this.model.getAllContentsByType([
      OntoUMLType.GENERALIZATION_SET_TYPE,
    ]) as IGeneralizationSet[];

    for (let i = 0; i < generalizationSets.length; i += 1) {
      const generalizationSet = generalizationSets[i];

      if (!generalizationSet.generalizations) continue;

      const classGeneralizations = (<IGeneralization[]>generalizationSet.generalizations).filter(
        (generalization: IGeneralization) =>
          generalization.specific.type === OntoUMLType.CLASS_TYPE,
      );
      const parent = classGeneralizations[0].general as IClass;
      const classes = classGeneralizations.map(
        (generalization: IGeneralization) => generalization.specific,
      );
      const classNodes = classes.map((classElement: IClass) => {
        const uri = getURI({ element: classElement, options });

        return namedNode(uri);
      });

      // check if has at least 2 classes to avoid insconsistence
      if (classNodes.length > 1) {
        // add disjoint
        if (generalizationSet.isDisjoint) {
          await writer.addQuad(
            writer.blank(namedNode('rdf:type'), namedNode('owl:AllDisjointClasses')),
            namedNode('owl:members'),
            writer.list(classNodes),
          );
        }

        // add complete
        if (generalizationSet.isComplete) {
          const parentUri = getURI({ element: parent, options });

          await writer.addQuad(
            namedNode(parentUri),
            namedNode('owl:equivalentClass'),
            writer.blank([
              {
                predicate: namedNode('rdf:type'),
                object: namedNode('owl:Class'),
              },
              {
                predicate: namedNode('owl:unionOf'),
                object: writer.list(classNodes),
              },
            ]),
          );
        }
      }
    }

    return true;
  }
}
