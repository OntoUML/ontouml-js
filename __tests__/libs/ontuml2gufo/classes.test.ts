import { ModelManager } from '@libs/model';
import { OntoUML2GUFO } from '@libs/ontuml2gufo';
import {
  genericExample1,
  modeExample1,
  modeExample2,
} from '@test-models/valids';
import { IPackage } from '@types';

async function transformOntoUML2GUFO(
  model: IPackage,
  format?: string,
): Promise<string> {
  const modelManager = new ModelManager(model);
  const service = new OntoUML2GUFO(modelManager);

  return await service.transformOntoUML2GUFO({
    baseIRI: 'https://example.com',
    format: format || 'N-Triple',
  });
}

it('should transform OntoUML generic model to gUFO', async () => {
  const result = await transformOntoUML2GUFO(genericExample1, 'Turtle');

  expect(result).toMatchSnapshot();
});

it('should transform OntoUML <<mode>> class as IntrinsicMode', async () => {
  const result = await transformOntoUML2GUFO(modeExample1);

  expect(result).toMatchSnapshot();

  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:Class>');
  expect(result).toContain(
    '<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:NamedIndividual>',
  );
  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdfs:label> "Headache" .');
  expect(result).toContain(
    '<:qJdeWA6AUB0UtAWm> <rdfs:subClassOf> <gufo:IntrinsicMode>',
  );
  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdf:type> <gufo:Kind>');
});

it('should transform OntoUML <<mode>> class as ExtrinsicMode', async () => {
  const result = await transformOntoUML2GUFO(modeExample2);

  expect(result).toMatchSnapshot();

  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:Class>');
  expect(result).toContain(
    '<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:NamedIndividual>',
  );
  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdfs:label> "Love" .');
  expect(result).toContain(
    '<:qJdeWA6AUB0UtAWm> <rdfs:subClassOf> <gufo:ExtrinsicMode>',
  );
  expect(result).toContain('<:qJdeWA6AUB0UtAWm> <rdf:type> <gufo:Kind>');
});
