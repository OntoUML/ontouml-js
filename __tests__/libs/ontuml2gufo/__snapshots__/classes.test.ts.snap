// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transform OntoUML <<mode>> class as ExtrinsicMode 1`] = `
"<https://example.com> <rdf:type> <owl:Ontology> .
<https://example.com> <owl:imports> <gufo:> .
[ <rdf:type> <owl:AllDisjointClasses> ] <owl:members> (<:Q5deWA6AUB0UtAWn> <:Q5deWA6AUB0UtAWo>) .
<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:Class> .
<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:NamedIndividual> .
<:qJdeWA6AUB0UtAWm> <rdfs:label> \\"Love\\" .
<:qJdeWA6AUB0UtAWm> <rdfs:subClassOf> <gufo:ExtrinsicMode> .
<:qJdeWA6AUB0UtAWm> <rdf:type> <gufo:Kind> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <owl:Class> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <owl:NamedIndividual> .
<:Q5deWA6AUB0UtAWn> <rdfs:label> \\"Person\\" .
<:Q5deWA6AUB0UtAWn> <rdfs:subClassOf> <gufo:FunctionalComplex> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <gufo:Kind> .
<:Q5deWA6AUB0UtAWo> <rdf:type> <owl:Class> .
<:Q5deWA6AUB0UtAWo> <rdf:type> <owl:NamedIndividual> .
<:Q5deWA6AUB0UtAWo> <rdfs:label> \\"Car\\" .
<:Q5deWA6AUB0UtAWo> <rdfs:subClassOf> <gufo:FunctionalComplex> .
<:Q5deWA6AUB0UtAWo> <rdf:type> <gufo:Kind> .
[ <rdf:type> <owl:AllDisjointClasses> ] <owl:members> (<:qJdeWA6AUB0UtAWm>) .
"
`;

exports[`should transform OntoUML <<mode>> class as IntrinsicMode 1`] = `
"<https://example.com> <rdf:type> <owl:Ontology> .
<https://example.com> <owl:imports> <gufo:> .
[ <rdf:type> <owl:AllDisjointClasses> ] <owl:members> (<:Q5deWA6AUB0UtAWn>) .
<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:Class> .
<:qJdeWA6AUB0UtAWm> <rdf:type> <owl:NamedIndividual> .
<:qJdeWA6AUB0UtAWm> <rdfs:label> \\"Headache\\" .
<:qJdeWA6AUB0UtAWm> <rdfs:subClassOf> <gufo:IntrinsicMode> .
<:qJdeWA6AUB0UtAWm> <rdf:type> <gufo:Kind> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <owl:Class> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <owl:NamedIndividual> .
<:Q5deWA6AUB0UtAWn> <rdfs:label> \\"Person\\" .
<:Q5deWA6AUB0UtAWn> <rdfs:subClassOf> <gufo:FunctionalComplex> .
<:Q5deWA6AUB0UtAWn> <rdf:type> <gufo:Kind> .
[ <rdf:type> <owl:AllDisjointClasses> ] <owl:members> (<:qJdeWA6AUB0UtAWm>) .
"
`;

exports[`should transform OntoUML generic model to gUFO 1`] = `
"@prefix : <https://example.com#>.
@prefix gufo: <http://purl.org/nemo/gufo#>.
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix owl: <http://www.w3.org/2002/07/owl#>.

<https://example.com> rdf:type owl:Ontology;
    owl:imports gufo:.
[ rdf:type owl:AllDisjointClasses ] owl:members (:2uxhBA6AUB0UtArb :j5phBA6AUB0UtAts).
:rw.hBA6AUB0UtArB rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Agent\\".
[ rdf:type owl:AllDisjointClasses ] owl:members (:j5phBA6AUB0UtAts :2uxhBA6AUB0UtArb).
:rw.hBA6AUB0UtArB rdfs:subClassOf gufo:Endurant;
    rdf:type gufo:Category;
    owl:equivalentClass [
  rdf:type owl:Class;
  owl:unionOf (:j5phBA6AUB0UtAts :2uxhBA6AUB0UtArb)
].
:2uxhBA6AUB0UtArb rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Person\\";
    rdfs:subClassOf :rw.hBA6AUB0UtArB, gufo:FunctionalComplex;
    rdf:type gufo:Kind.
:j5phBA6AUB0UtAts rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Organization\\";
    rdfs:subClassOf :rw.hBA6AUB0UtArB, gufo:FunctionalComplex;
    rdf:type gufo:Kind.
:_xThBA6AUB0UtAvg rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"School\\";
    rdfs:subClassOf :j5phBA6AUB0UtAts.
[ rdf:type owl:AllDisjointClasses ] owl:members (:JLnhBA6AUB0UtAwO).
:_xThBA6AUB0UtAvg rdf:type gufo:SubKind.
:GDbhBA6AUB0UtAv1 rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Student\\";
    rdfs:subClassOf :2uxhBA6AUB0UtArb;
    rdf:type gufo:Role.
:JLnhBA6AUB0UtAwO rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Enrollment\\";
    rdfs:subClassOf gufo:Relator;
    rdf:type gufo:Kind.
:wukRBA6AUB0UtAxW rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Enrollment Date\\".
[ rdf:type owl:AllDisjointClasses ] owl:members (:wukRBA6AUB0UtAxW).
:wukRBA6AUB0UtAxW rdfs:subClassOf gufo:Quality;
    rdf:type gufo:Kind.
:h_Y9hA6GAqACnA1t rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Sophomore\\";
    rdfs:subClassOf :GDbhBA6AUB0UtAv1;
    rdf:type gufo:Role.
:cea9hA6GAqACnA2B rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Privately Enrolled\\";
    rdfs:subClassOf :GDbhBA6AUB0UtAv1;
    rdf:type gufo:Role.
:6XxdhA6GAqACnA0J rdf:type owl:Class, owl:NamedIndividual;
    rdfs:label \\"Department\\";
    rdfs:subClassOf :j5phBA6AUB0UtAts;
    rdf:type gufo:SubKind.
"
`;
